From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Thu, 4 Jun 2026 00:56:32 +0800
Subject: [PATCH] Add config to enable trigger command


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 5fda3083f73744b0df3aab47b3c378dba41b904c..82af03649c7537274aa54ed942e4aae589b97888 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -274,7 +274,11 @@ public class Commands {
         // Luminol end - Add a config to enable tick command
         TimeCommand.register(this.dispatcher, context);
         TitleCommand.register(this.dispatcher, context);
-        //TriggerCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        // Lophine start - Add a config to enable trigger command
+        if (fun.bm.lophine.config.modules.experiment.CommandConfig.trigger) {
+            TriggerCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        }
+        // Lophine end - Add a config to enable trigger command
         if (me.earthme.luminol.config.modules.experiment.CommandConfig.waypointsAndWaypointCommand) WaypointCommand.register(this.dispatcher, context); // Folia - region threading - TODO later // Luminol - Restore waypoints
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
